### DISCUSSION TITLE
Use private nuget repository

### DISCUSSION THREAD

**Discussion by bluecitylights** [renovatebot/renovate#8713] [Request Help]
My project contains a nuget.config file that uses a private repository and has the password configured as an environment variable.

```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <clear />
    <add key="api.nuget.org" value="https://api.nuget.org/v3/index.json" /> 
    <add key="platform-lib" value="https://gitlab.com/api/v4/projects/xxxx/packages/nuget/index.json" />
  </packageSources>
  <packageSourceCredentials>
    <platform-lib>
        <add key="Username" value="Platform_Packages" />
        <add key="ClearTextPassword" value="%PLATFORM_PACKAGES_PASSWORD%" />
      </platform-lib>

  </packageSourceCredentials>
</configuration>
```

When I onboard Renovate, the MR mentions : "Please correct - or verify that you can safely ignore - these lookup failures before you merge this MR.

Failed to look up dependency PlatformLib"

I tried adding this repo to the hostrules like this:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "packageRules": [
    {
      "datasources": [
        "nuget"
      ],
      "registryUrls": [
        "https://api.nuget.org/v3/index.json",
        "https://gitlab.com/api/v4/projects/xxx/packages/nuget/index.json"
      ]
    }
  ],
  "hostRules": [
    {
      "hostType": "nuget",
      "baseUrl": "https://gitlab.com/api/v4/projects/xxx/packages/nuget",
      "username": "Platform_Packages",
      "password": "$PLATFORM_PACKAGES_PASSWORD"
    }
  ],

  "automerge": true,
  "major": {
    "automerge": false
  }
}
```

However, it seems the bot is not picking upgrading the package to a higher version.
Can we use hostrules and nuget.config simulatanuously?

**Comment by rarkins** [comment]
Did you put in the actual password here? `$PLATFORM_PACKAGES_PASSWORD` or is that the string you put?

Renovate does not do environment substitution in JSON configs.

You could do something similar in the `config.js` like:

```js
  "hostRules": [
    {
      "hostType": "nuget",
      "baseUrl": "https://gitlab.com/api/v4/projects/xxx/packages/nuget",
      "username": "Platform_Packages",
      "password": process.env.PLATFORM_PACKAGES_PASSWORD
    }
  ],
```

**Reply by bluecitylights** [reply]
I will give it a try. Actually we have removed nuget.config completely, and are just generating it, before running build.

**Reply by bluecitylights** [reply]
Like this I got it working now. So hostrules in config.js like you described, and no nuget.config in the repo.
Thanks

Edit: it actually works even thre is a nuget.config in the repo.

**Reply by bluecitylights** [reply]
@rarkins One more question can i add all settings either in config.json or in renovate.json? So config.json will be used as the default for repos? eg automerge?

**Reply by rarkins** [reply]
Admin-only options can only be put into the bot config (config.js or config.json).

User config options can be put in either, but it's recommended to usually put it in the repo config instead of bot config because otherwise they're too "invisible" to users. One exception to this is hostRules with authentication, because they contain secrets.

If you want to achieve DRY then the recommended solution is to use renovate presets to centralize config.
